doc/source/rllib/external-envs.rst

```diff
@@ -0,0 +1,142 @@
+.. include:: /_includes/rllib/we_are_hiring.rst
+
+.. include:: /_includes/rllib/new_api_stack.rst
+
+.. _rllib-external-env-setups-doc:
+
+
+External Environments and Applications
+--------------------------------------
+
+In many situations, it does not make sense for an RL environment to be "stepped" by RLlib.
+For example, if we train one or more policies inside a complex simulator, for example, a game engine
+or a robotics simulation, it would be more natural and user friendly to flip this setup around
+and - instead of RLlib "stepping" the env - allow the simulations and the agents to fully control
+their own stepping. An external RLlib-powered service would be available for either querying for
+individual actions or for accepting batched sample data. The service would cover the task
+of training the policies, but would not pose any restrictions on when and how often
+per second the simulation should step.
+
+.. figure:: images/envs/external_env_setup_client_inference.svg
+    :width: 600
+    **External application with client-side inference**: An external simulator (for example a game engine)
+    connects to RLlib, which runs as a server through a tcp-cabable, custom EnvRunner.
+    The simulator sends batches of data from time to time to the server and in turn receives weights updates.
+    For better performance, actions are computed locally on the client side.
+
+.. todo (sven): show new image here with UE5
+.. .. figure:: images/rllib-training-inside-a-unity3d-env.png
+.. scale: 75 %
+..    A Unity3D soccer game being learnt by RLlib via the ExternalEnv API.
+
+RLlib provides an `external messaging protocol <https://github.com/ray-project/ray/blob/master/rllib/env/utils/external_env_protocol.py>`__
+for this purpose as well as the option to customize your :py:class:`~ray.rllib.env.env_runner.EnvRunner` class used for collecting
+training data. An example, `tcp-based EnvRunner implementation is available here <https://github.com/ray-project/ray/blob/master/rllib/examples/envs/env_connecting_to_rllib_w_tcp_client.py>`__,
+together with a dummy (CartPole) client providing a template for any external application or simulator.
+
+.. note::
+    External application support is still work-in-progress on RLlib's new API stack. The Ray team
+    is working on more examples for custom EnvRunner implementations (besides
+    `the already available tcp-based one <https://github.com/ray-project/ray/blob/master/rllib/env/tcp_client_inference_env_runner.py>`__)
+    as well as client-side, non-python RLlib-adapters, for example for popular game engines and other
+    simulation software.
+
+
```

Comment:
Still working on the rest of this very file ... Everything else (all other files) is ready, though.